### PR TITLE
SecondLevel QueryCache causes invalid cast

### LINKED ISSUE #1385: SecondLevelCache CreateSQLQuery().UniqueResult<int>() throws Exception Specified cast is not valid.
Hi

Please check if this is a bug i get `Exception: System.InvalidCastException : Specified cast is not valid`.

For test case these two flags must be enabled:
```
 cache.use_second_level_cache=true 
 cache.use_query_cache=true 
```

TestCase:

```c#
[Test]
public void QueryCacheWithScalarFunctionThatReturnsInt()
{
	using (ISession s = OpenSession())
	{
		int result =  s
			.CreateSQLQuery("SELECT [dbo].[TestQueryCache] ()")
			.SetCacheable(true)
			.UniqueResult<int>();

		Assert.IsNotNull(result);
	}
}
```

Mssql scalar function create:

```sql
USE [nhibernate]
GO

/****** Object:  UserDefinedFunction [dbo].[TestQueryCache]    Script Date: 17.10.2017 14:11:13 ******/
SET ANSI_NULLS ON
GO

SET QUOTED_IDENTIFIER ON
GO

-- =============================================
-- Author:		<Author,,Name>
-- Create date: <Create Date, ,>
-- Description:	<Description, ,>
-- =============================================
CREATE FUNCTION [dbo].[TestQueryCache] 
()
RETURNS INT
AS
BEGIN
	
	RETURN DATEPART(YEAR, GETUTCDATE()) * 100 + DATEPART(MONTH, GETUTCDATE());

END

GO
```

### PR DESCRIPTION
**Pull request by fredericDelaporte**
* Fix #1385

As said in #1385, I am not sure we should really try fixing this. Maybe we should instead detect the case (detect attempt to cache a query requiring types auto-discovery) and yield a proper error, instead of letting it fails with an invalid cast.

### CODE REVIEW

**Comment by fredericDelaporte** [comment]
> Why not use [Theory] instead of ignoring bunch of tests?

You mean, by example changing:
```c#
[Test]
public void LoadElementCollectionWithCustomLoader()
{
	if (WithQueryCache)
	{
		Assert.Ignore("This test does not use a cacheable query.");
	}
	string[] routes = CreateRoutes();
	Country country = CreateCountry(routes);
	Save(country);
	using (ISession session = OpenSession())
	{
		var c = session.Get<Country>(country.Code);
		Assert.That(c, Is.Not.Null, "country");
		Assert.That(c.Routes, Is.EquivalentTo(routes), "country.Routes");
	}
}
```
To:
```c#
[Theory]
public void LoadElementCollectionWithCustomLoader()
{
	Assume.That(WithQueryCache, Is.False, "This test does not use a cacheable query.");
	string[] routes = CreateRoutes();
	Country country = CreateCountry(routes);
	Save(country);
	using (ISession session = OpenSession())
	{
		var c = session.Get<Country>(country.Code);
		Assert.That(c, Is.Not.Null, "country");
		Assert.That(c.Routes, Is.EquivalentTo(routes), "country.Routes");
	}
}
```
Why not.

**Comment by hazzik** [comment]
No, this is not what I meant.

**Comment by hazzik** [comment]
I meant using `[Theory]` instead of `WithQueryCache` property

**Comment by hazzik** [comment]
Ah, I did not see that `WithQueryCache` is used in `SetUp`.

**Comment by fredericDelaporte** [comment]
> UPD2: Ah, I did not see that `WithQueryCache` is used in `SetUp`.

It could be removed from there since the query cache is activated in our current test configuration, and since it does not really need to be de-activated because it is also handled on a query per query basis.

But then I would have to do a bit more changes for tests using `LoadCountryWithNativeSQL`. That was for minimizing changes to those tests that I have chosen using an overridable property.

Still switching from `Ignore` to `Assume` is less code noisy.

**Comment by hazzik** [comment]
But then you don't need to use `[Theory]` flag. It should work with `[Test]`, shouldn't it?

**Comment by fredericDelaporte** [comment]
`Assume` is meant to be used with `Theory`, not sure `Test` accepts it. Going to check.

File: src/NHibernate/Transform/CacheableResultTransformer.cs [hunk 1]
```diff
@@ -152,13 +172,24 @@ private CacheableResultTransformer(bool[] includeInTuple, bool[] includeInTransf
...
 		public object TransformTuple(object[] tuple, string[] aliases)
 		{
-			if (aliases != null && aliases.Length != _tupleLength)
+			if (_includeInTuple == null)
+				throw new InvalidOperationException("This transformer is not initialized");
+			if (aliases != null && aliases.Length != _includeInTuple.Length)
 			{
 				throw new InvalidOperationException(
-					"aliases expected length is " + _tupleLength +
+					"aliases expected length is " + _includeInTuple.Length +
```
Comment (fredericDelaporte) [inline src/NHibernate/Transform/CacheableResultTransformer.cs]: This change is questioning. Tests succeeds, but why should it be so? Moreover it appears here that `aliases` are only used for causing this failure. The actual implementation does not need them. Maybe this check should be removed instead.

File: src/NHibernate/Transform/CacheableResultTransformer.cs [hunk 2]
```diff
@@ -152,22 +172,23 @@ private CacheableResultTransformer(bool[] includeInTuple, bool[] includeInTransf
...
 		{
-			if (aliases != null && aliases.Length != _tupleLength)
-			{
-				throw new InvalidOperationException(
-					"aliases expected length is " + _tupleLength +
-					"; actual length is " + aliases.Length);
-			}
-			// really more correct to pass index( aliases.getClass(), aliases )
-			// as the 2nd arg to the following statement;
-			// passing null instead because it ends up being ignored.
```
Comment (fredericDelaporte) [inline src/NHibernate/Transform/CacheableResultTransformer.cs]: Test and comment on `aliases` here were quite obsolete, and the test was causing the auto-discovered cached query to fail. `aliases` are not used by the actual implementation, and the suggested one use a method which seems suppressed.

**Comment by fredericDelaporte** [comment]
@gliljas, @hazzik, do you think #1385 (fixed by this PR) should be in 5.1?

**Comment by fredericDelaporte** [comment]
Rebased.

**Comment by fredericDelaporte** [comment]
There are some people seemingly strongly awaiting this fix (see #1385), so for them it would be great to have it in 5.1.

Now, this fix is still a bit hackish in my view, although I do not think it may creates new issues.

**Comment by hazzik** [comment]
This goes to 5.1 and everything else are going to next release.

**Pull request merged**